backend/pvm.pm

```diff
@@ -0,0 +1,151 @@
+
+    my $id = substr(qx/pvmctl lpar list -d LogicalPartition.id --where LogicalPartition.name=$vars->{LPAR}/, 3);
+    chomp($id);
+    $vars->{LPARID} = $id;
+    diag "vars iso $vars->{ISO}";
+    diag "LPAR is $vars->{LPAR}, ID is $vars->{LPARID}";
+    my $os      = $vars->{DISTRI};
+    my $version = $vars->{VERSION};
+    if ($vars->{DISTRI} =~ /opensuse/) {
+        $os      = "os";
+        $version = "tw";
+    }
+    $vars->{VIOISO} = "$os-$version-$vars->{FLAVOR}-$vars->{ARCH}-$vars->{BUILD}.iso";
+    my $iso = $vars->{VIOISO};
+    #unfortunately NovaLink can't take files longer than 38 chars. So we need to shorten ISO name here
+    #we copy isos from nfs mount on VIO side to VMLibrary
+    my $source_iso = $vars->{ISO};
+    $source_iso =~ s/var\/lib\/openqa\/share\/factory\///g;
+    diag "source_iso: $source_iso, iso: $iso";
+    my $iso_present = qx/pvmctl media list -d VirtualOpticalMedia.media_name --where VirtualOpticalMedia.name=$iso/;
+    if ($iso_present =~ /$iso/) {
+        runcmd($pvmctl, "scsi", "create", "--type", "vopt", "--lpar", "name=" . $vars->{LPAR}, "--stor-id", $iso);
+    }
+    else {
+        #copy over from nfs to VMLibrary
+        my $cmd = "sudo viosvrcmd --id 1 -r -c\" cp $source_iso /var/vio/VMLibrary/$iso\"";
+        runcmd($cmd);
+        $cmd = "sudo viosvrcmd --id 1 -r -c\" chown padmin:staff /var/vio/VMLibrary/$iso\"";
+        runcmd($cmd);
+        runcmd($pvmctl, "scsi", "create", "--type", "vopt", "--lpar", "name=" . $vars->{LPAR}, "--stor-id", $iso);
+    }
+
+    for my $i (1 .. $vars->{NUMDISKS}) {
+        runcmd($pvmctl, "lv", "create", "--name", $vars->{LPAR} . "_" . $i, "--size", $vars->{HDDSIZEGB});
+        runcmd($pvmctl, "scsi", "create", "--vg", "name=rootvg", "--type", "lv", "--lpar", "name=" . $vars->{LPAR}, "--stor-id", $vars->{LPAR} . "_" . $i);
+    }
+
+    runcmd($pvmctl, "eth", "create", "--pvid", $vars->{NICVLAN}, "--vswitch", $vars->{VSWITCH}, "-p", "id=" . $vars->{LPARID});
+    bmwqemu::save_vars();
+    powercontrol('on');
+    attach_console;
+
+    my $vnc = $testapi::distri->add_console(
+        'sut',
+        'vnc-base',
+        {
+            hostname => 'localhost',
+            port     => $vars->{VNC}});
+    $vnc->backend($self);
+    $self->select_console({testapi_console => 'sut'});
+}
+
+sub do_stop_vm {
+    my $self   = shift;
+    my $vars   = \%bmwqemu::vars;
+    my $pvmctl = $self->{pvmctl};
+    my $cmd;
+    powercontrol('off');
+    runcmd("rmvterm", "--id", $vars->{LPARID});
+    for my $i (1 .. $vars->{NUMDISKS}) {
+        my $disk = $vars->{LPAR} . "_" . $i;
+        $cmd = "$pvmctl scsi delete --type lv --lpar name=$vars->{LPAR} --stor-id=$disk";
+        runcmd($cmd);
+        $cmd = "sudo viosvrcmd --id 1 -c \"rmbdsp -sp rootvg -bd $disk\"";
+        runcmd($cmd);
+    }
+    $cmd = "$pvmctl scsi delete --type vopt --lpar name=$vars->{LPAR} --stor-id=$vars->{VIOISO}";
```

Comment:
what bothers me here is that you duplicate this code - and even inconsistently. Here you use a string and in the create case you use an array. Both share a lot of paramters, so possibly they both should inherit a common array.